File: MergeSortedLinkedList/MergeSortedLinkedList.cpp
// MergeSortedLinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

std::ostream& operator <<(std::ostream& st, ListNode* pNode)
{
    if (pNode == nullptr)
        return st;

    st << pNode->data;
    if (pNode->next == nullptr)
        st << std::endl;
    else
    {
        st << ", " << pNode->next;
    }

    return st;
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

ListNode* MergeTwoSortedList(ListNode* firstList, ListNode* secondList)
{
    if (firstList == nullptr)
        return secondList;
    if (secondList == nullptr)
        return firstList;

    ListNode* newHead = firstList;
    if (newHead->data > secondList->data)
    {
        newHead = secondList;
        secondList = secondList->next;
    }
    else
        firstList = firstList->next;

    ListNode* newCurrent = newHead;

    while (firstList != nullptr && secondList != nullptr)
    {
        if (firstList->data < secondList->data)
        {
            newCurrent->next = firstList;
            firstList = firstList->next;
        }
        else
        {
            newCurrent->next = secondList;
            secondList = secondList->next;
        }
        newCurrent = newCurrent->next;
    }

    if (firstList != nullptr)
    {
        newCurrent->next = firstList;
    }

    if (secondList != nullptr)
    {
        newCurrent->next = secondList;
    }

    return newHead;
}

int main()
{
    while (true)
    {
        int countA = 0, countB = 0;
        std::cout << "Number of nodes in listA and listB: ";
        std::cin >> countA >> countB;
        if (countA == 0 && countB == 0)
            break;

        ListNode* pAHead = nullptr, *pBHead = nullptr;
        ListNode* pATail = GetListFromInput(&pAHead, countA);
        ListNode* pBTail = GetListFromInput(&pBHead, countB);

        ListNode* pMerged = MergeTwoSortedList(pAHead, pBHead);

        std::cout << "Merged: ";
        std::cout << pMerged;
        FreeList(pMerged);
    }
}

File: Go/MergeSortedLinkedList.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) *ListNode {
	if len(v) == 0 {
		return nil
	}

	if len(v) == 1 {
		return &ListNode{v[0], nil}
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead
}

func (node *ListNode) Print() {
	if node == nil {
		fmt.Println("")
	}

	fmt.Print(node.data)
	node = node.next
	for node != nil {
		fmt.Printf(", %d", node.data)
		node = node.next
	}
	fmt.Println("")
}

func MergeTwoSortedList(firstList, secondList *ListNode) *ListNode {
	if firstList == nil {
		return secondList
	}
	if secondList == nil {
		return firstList
	}

	pHead := firstList
	if firstList.data >= secondList.data {
		pHead = secondList
		secondList = secondList.next
	} else {
		firstList = firstList.next
	}

	pCur := pHead

	for firstList != nil && secondList != nil {
		if firstList.data < secondList.data {
			pCur.next = firstList
			firstList = firstList.next
		} else {
			pCur.next = secondList
			secondList = secondList.next
		}
		pCur = pCur.next
	}

	if firstList != nil {
		pCur.next = firstList
	} else {
		pCur.next = secondList
	}

	return pHead
}

func main() {
	for {
		fmt.Print("Number elements in list1 and list2: ")
		var countA, countB int
		fmt.Scan(&countA, &countB)
		if countA == 0 && countB == 0 {
			break
		}

		vec := GetArrayFromInput(countA, "Please enter the nodes in list1: ")
		listA := GetListFromArray(vec)

		vec = GetArrayFromInput(countB, "Please enter the nodes in list2: ")
		listB := GetListFromArray(vec)

		pHead := MergeTwoSortedList(listA, listB)
		fmt.Print("The merged list: ")
		pHead.Print()
	}
}
